### ISSUE TITLE
Not bound to wrappedComponent

### ISSUE THREAD

**Issue by ZeroCho** [kodefox/class-autobind#2]
Ok, first of all, below is my codes

```
import ...
class LoginForm extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = { message: ''};
    autobind(this);
  }
  handleSubmit() {
    this.setState({ message: ''}); // This is where the error happens
    ...
  }
  render() { ... }
}
```

And the error happens when I submit the form.
`Uncaught TypeError: Cannot read property 'setState' of null`

So I inspected your code and console-logged followings

```
constructor(props) {
    super(props);
    this.state = { message: '' };
    console.log(this);
    console.log(this.constructor);
    console.log(this.constructor.prototype);
    console.log(Object.getOwnPropertyNames(this.constructor.prototype));
    autobind(this);
  }
```

And each said like below
`LoginForm {props: Object, context: Object, refs: Object, updater: Object, state: Object}`

```
function WrappedComponent() {
            _classCallCheck(this, WrappedComponent);
            return _possibleConstructorReturn(this, _Component.apply(this, arguments));
        }
```

`LoginForm {}`
`["constructor", "render"] // handleSubmit method disappeared` 

I expected that the result of `console.log(this.constructor);` to be `function Login()`, but actually it shows `function wrappedComponent()` and autobind function cannot bind this with methods.

**Comment by ZeroCho** [comment]
Oh, I found that it is wrapped with react-css-modules. I guess that this package cannot be used with react-css-modules

**Comment by sstur** [comment]
Hi ZeroCho.

I will try to help you solve this. I do not think it's related to your use of react-css-modules. I think there is something else happening.

I suspect that somewhere in your render method you are passing `this.handleSubmit` to a child component. Can you show that code?

As far as the console.log'ing that output seems OK to me. The bound `handleSubmit` would NOT be present on `this.constructor.prototype` but should be present on `this` directly. So if you were to do: `console.log(Object.getOwnPropertyNames(this))` you should see it, right?

Edit: I see what you were trying to log. The `handleSubmit` _before_ binding should be present on the prototype. I'm not sure why it's not. What happens if you log `typeof this.constructor.prototype.handleSubmit`? Actually `this.constructor.prototype.foo` should be the same as `super.foo` in ES6. After the `autobind` call, can you log `typeof this.handleSubmit` and also log `this.handleSubmit === super.handleSubmit`?

**Comment by ZeroCho** [comment]
Well, **Login** component is a child component of **MainLoginPage** component. And,  **this.handleSubmit** isn't passed to a child component **but** uses one of props which is transferred by parent MainLoginPage component. MainLoginPage is connected with react-redux.

below is the full code.

```
import React, { Component, PropTypes } from "react";
import CSSModules from 'react-css-modules';
import autobind from 'class-autobind';
import styles from "../css/shared/Form.css";

class LoginForm extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      message: '',
    };
    autobind(this);
    console.log(this);
    console.log(Object.getOwnPropertyNames(this));
    console.log(this.constructor);
    console.log(this.constructor.prototype);
    console.log(this.__proto__);
    console.log(Object.getOwnPropertyNames(this.__proto__));
    console.log(typeof this.handleSubmit);
    console.log(this.handleSubmit === super.handleSubmit);
    console.log('super:', typeof super.handleSubmit); // Should be "function"
    console.log('this:', typeof this.handleSubmit); // Should be "function"
    console.log('is same:', this.handleSubmit === super.handleSubmit); // Should be false
  }

  handleForgotPassword(e) {
    e.preventDefault();
    this.setState({ message: '' });
    const { forgotPassword } = this.props;
    const email = this.refs.email.value && this.refs.email.value.trim();
    if (!email) {
      return this.setState({ message: '이메일 칸에 이메일을 입력한 후 버튼을 누르세요.' });
    }
    return forgotPassword({ email });
  }

  handleSubmit(e) {
    e.preventDefault();
    this.setState({ message: '' });
    const email = this.refs.email.value;
    const password = this.refs.password.value;
    if (!email || !email.trim()) {
      return this.setState({ message: '이메일을 입력하세요' });
    }
    if (!password || password.length < 8) {
      return this.setState({ message: '비밀번호는 8자리 이상이어야 합니다.' });
    }
    return this.props.onSubmitLoginForm({ email, password });
  }

  render() {
    const { isFetchingData, isFindingPass, serverError, goBack, message } = this.props;
    return (
      <form styleName="form" onSubmit={this.handleSubmit}>
        <div>
          <label htmlFor="email" styleName="fieldLabel">이메일</label>
          <input type="email" ref="email" name="email" styleName="fieldInput" />
        </div>
        <div>
          <label htmlFor="password" styleName="fieldLabel">비밀번호</label>
          <input type="password" ref="password" name="password" styleName="fieldInput" />
        </div>
        <h4>상단 메뉴에 있는 SNS로도 간단하게 로그인 할 수 있어요!</h4>
        <div className="submit-menu">
          <button type="submit">
            {isFetchingData ? '로그인 중...' : '로그인'}
          </button>
          <button type="reset" onClick={goBack}>취소</button>
          <button onClick={this.handleForgotPassword}>
            {isFindingPass ? '메일 전송 중' : '비밀번호를 잊어버렸어요'}
          </button>
        </div>
        {message ? <div>{message}</div> :
          <div styleName="errorMessage">{serverError || this.state.message}</div>}
      </form>
    );
  }
}

LoginForm.propTypes = {
  forgotPassword: PropTypes.func.isRequired,
  onSubmitLoginForm: PropTypes.func.isRequired,
  isFetchingData: PropTypes.bool.isRequired,
  isFindingPass: PropTypes.bool,
  serverError: PropTypes.string,
  goBack: PropTypes.func.isRequired,
  message: PropTypes.string,
};

LoginForm.contextTypes = {
  router: PropTypes.object.isRequired,
};

export default CSSModules(LoginForm, styles);
```

I'll try to reproduce the same error and tell you the result of console.log. Please, wait a minute.

**Comment by sstur** [comment]
Everything in your code above looks good. I'm not sure why it is throwing that error.

**Comment by ZeroCho** [comment]
Console.log results **without** autobind

```
console.log(this);
LoginForm {props: Object, context: Object, refs: Object, updater: Object, state: Object} // I can see handleSubmit method. it's undefined because it's not bound yet, right?
```

```
console.log(Object.getOwnPropertyNames(this));
["props", "context", "refs", "updater", "state"]
```

```
console.log(this.constructor);
function WrappedComponent() {
            _classCallCheck(this, WrappedComponent);

            return _possibleConstructorReturn(this, _Component.apply(this, arguments));
        }
```

```
console.log(this.constructor.prototype);
LoginForm {} // no handleSubmit
```

```
console.log(Object.getOwnPropertyNames(this.constructor.prototype));
["constructor", "render"]
```

```
console.log(typeof this.handleSubmit);
function
```

```
console.log(this.handleSubmit === super.handleSubmit);
false
```

```
console.log(this.__proto__);
WrappedComponent {} // no handleSubmit
```

```
console.log(Object.getOwnPropertyNames(this.__proto__));
["constructor", "render", "componentDidMount", "componentWillUnmount"]
```

**Comment by ZeroCho** [comment]
Console.log results **with** autobind

```
console.log(this);
LoginForm {props: Object, context: Object, refs: Object, updater: Object, state: Object} //No handleSubmit method
```

```
console.log(Object.getOwnPropertyNames(this));
["props", "context", "refs", "updater", "state", "constructor"]
```

```
console.log(this.constructor);
function WrappedComponent() {
            _classCallCheck(this, WrappedComponent);

            return _possibleConstructorReturn(this, _Component.apply(this, arguments));
        }
```

```
console.log(this.constructor.prototype);
undefined
```

```
console.log(Object.getOwnPropertyNames(this.constructor.prototype));
TypeError: Cannot convert undefine or null to object // because this.contstructor.prototype is undefined
```

```
console.log(typeof this.handleSubmit);
function
```

```
console.log(this.handleSubmit === super.handleSubmit);
false
```

```
console.log(this.__proto__);
WrappedComponent {} // no handleSubmit, either
```

```
console.log(Object.getOwnPropertyNames(this.__proto__));
["constructor", "render", "componentDidMount", "componentWillUnmount"]
```

**Comment by sstur** [comment]
Oops, that last one should be `this.handleSubmit === super.handleSubmit` (note the triple equal).

**Comment by sstur** [comment]
Your logging indicates that we cannot rely on `this.constructor` because something is replacing it. I'd suggest avoiding use of that syntax anyway.

Could you do your logging again except completely remove the `console.log(this.constructor)` part and replace all the `this.constructor.prototype` with `this.__proto__`?

You can edit the comments above rather than post a new set of logs.

I'm a little confused why `super.handleSubmit` would be undefined.

**Comment by sstur** [comment]
To be honest, I think that all your code is correct with regards to the constructor and the logging. I think your problem lies somewhere else. But I can't tell from this logging what the issue is.

**Comment by ZeroCho** [comment]
I added two more console.log results(**proto**) and edited ===

**Comment by sstur** [comment]
OK. We're probably logging more than we need. But this part indicates to me that autobind is working:

```
console.log(typeof this.handleSubmit);
function
```

```
console.log(this.handleSubmit === super.handleSubmit);
false
```

Now that I think about it, all we really need to log is this:

```
console.log('super:', typeof super.handleSubmit); // Should be "function"
console.log('this:', typeof this.handleSubmit); // Should be "function"
console.log('is same:', this.handleSubmit === super.handleSubmit); // Should be false
```

If all those are as expected, then the problem is somewhere else. But I'm not sure where.

**Comment by ZeroCho** [comment]
super: undefined
this: function
is same: false

super.handleSubmit is undefined... and I uploaded full LoginForm.js in the comment above

**Comment by sstur** [comment]
Oh, that's very strange. Something is messing with your `__proto__` chain and/or `prototype.constructor`. I'd have to see the rest of your code to understand what is going on.

**Comment by ZeroCho** [comment]
Which part? Other than LoginForm.js? I uploaded full LoginForm.js above

**Comment by ZeroCho** [comment]
It's a MainLoginPage.js which uses LoginForm.js

```
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

import LoginForm from '../components/LoginForm.js';
import { attemptLogin, attemptFindPassword } from '../actions/UserActions';

class MainLoginPage extends Component {
  constructor(props) {
    super(props);
    this.onSubmitLoginForm = this.onSubmitLoginForm.bind(this);
  }

  onSubmitLoginForm(formData) {
    const { login } = this.props;
    login(formData.email, formData.password);
  }

  render() {
    const { findPass, userAuthSession } = this.props;
    if (userAuthSession.isLoggedIn) {
      this.context.router.goBack();
    }
    return (
      <div id="login">
        <h1>로그인</h1>
        <LoginForm
          onSubmitLoginForm={this.onSubmitLoginForm}
          forgotPassword={(data) => findPass(data)}
          isFetchingData={userAuthSession.fetchingAuthUpdate}
          isFindingPass={userAuthSession.fetchingFindUpdate}
          serverError={userAuthSession.error}
          message={userAuthSession.message}
          goBack={this.context.router.goBack}
        />
      </div>
    );
  }
}

MainLoginPage.propTypes = {
  findPass: PropTypes.func.isRequired,
  login: PropTypes.func.isRequired,
  userAuthSession: PropTypes.object.isRequired,
};

MainLoginPage.contextTypes = {
  router: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    userAuthSession: state.userAuthSession,
  };
}

export default connect(mapStateToProps, {
  login: attemptLogin, findPass: attemptFindPassword,
})(MainLoginPage);
```

**Comment by ZeroCho** [comment]
Without react-css-modules, it works fine

**Comment by sstur** [comment]
Your code looks absolutely perfect to me. I see no reason why there should be any issues! I'll try to reproduce this locally.

**Comment by sstur** [comment]
OK, I see. This is most certainly related to `react-css-modules`. I looked into their source code and they do indeed mess with the prototype chain of your component. Specifically, they extend it and override the `render` method. I can work around this in `autobind` by accepting an optional second parameter. Something like: `autobind(this, LoginForm.prototype);`

It would be a small change on my part and it would allow you to use `autobind` with `react-css-modules`.

However, I think the approach taken by `react-css-modules` is generally a poor practice. It breaks some assumptions that developers would make, namely that `this` within the constructor of a React component actually points to the instance being rendered. Furthermore, they mutate the props and generally do some hacky stuff.

I'll open an issue on that project and suggest a better approach.

**Comment by ZeroCho** [comment]
Thanks a lot!

**Comment by sstur** [comment]
I've implemented the workaround in v0.1.4 which is now published on npm. You can now use `autobind(this, LoginForm.prototype)` and it should work.

Thanks for reporting this!

**Comment by ZeroCho** [comment]
It works like magic! Thank you again!
